x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java

```diff
@@ -44,11 +61,157 @@ public LogicalPlan visitEventQuery(EqlBaseParser.EventQueryContext ctx) {
+                queries.add(joinTerm);
+            }
+        }
+
+        return new Join(source(ctx), queries, until);
+    }
+
+    public KeyedFilter visitJoinTerm(JoinTermContext ctx, List<Expression> joinKeys) {
+        List<Expression> keys = CollectionUtils.combine(joinKeys, visitJoinKeys(ctx.by));
+        return new KeyedFilter(source(ctx), visitEventQuery(ctx.subquery().eventQuery()), keys);
+    }
+
+    @Override
+    public Sequence visitSequence(SequenceContext ctx) {
+        List<Expression> parentJoinKeys = visitJoinKeys(ctx.by);
+
+        TimeValue maxSpan = visitSequenceParams(ctx.sequenceParams());
+
+        LogicalPlan until;
+
+        if (ctx.until != null) {
+            until = visitSequenceTerm(ctx.until, parentJoinKeys);
+        } else {
+            // no until declared means the condition never gets executed and thus folds to false
+            until = new Filter(source(ctx), RELATION, new Literal(source(ctx), Boolean.FALSE, DataTypes.BOOLEAN));
+        }
+
+        int numberOfKeys = -1;
+        List<LogicalPlan> queries = new ArrayList<>(ctx.sequenceTerm().size());
+
+        for (SequenceTermContext sequenceTermCtx : ctx.sequenceTerm()) {
+            KeyedFilter sequenceTerm = visitSequenceTerm(sequenceTermCtx, parentJoinKeys);
+            int keySize = sequenceTerm.keys().size();
+            if (numberOfKeys < 0) {
+                numberOfKeys = keySize;
+            } else {
+                if (numberOfKeys != keySize) {
+                    Source src = source(sequenceTermCtx.by != null ? sequenceTermCtx.by : sequenceTermCtx);
+                    int expected = numberOfKeys - parentJoinKeys.size();
+                    int found = keySize - parentJoinKeys.size();
+                    throw new ParsingException(src, "Inconsistent number of join keys specified; expected [{}] but found [{}]", expected,
+                            found);
+                }
+                queries.add(sequenceTerm);
+            }
+        }
+
+        return new Sequence(source(ctx), queries, until, maxSpan);
+    }
+
+    public KeyedFilter visitSequenceTerm(SequenceTermContext ctx, List<Expression> joinKeys) {
+        if (ctx.FORK() != null) {
+            throw new ParsingException(source(ctx.FORK()), "sequence fork is unsupported");
+        }
+
+        List<Expression> keys = CollectionUtils.combine(joinKeys, visitJoinKeys(ctx.by));
+        return new KeyedFilter(source(ctx), visitEventQuery(ctx.subquery().eventQuery()), keys);
+    }
+
+    @Override
+    public TimeValue visitSequenceParams(SequenceParamsContext ctx) {
+        if (ctx == null) {
+            return TimeValue.MINUS_ONE;
+        }
+
+        NumberContext numberCtx = ctx.timeUnit().number();
+        if (numberCtx instanceof IntegerLiteralContext) {
+            Number number = (Number) visitIntegerLiteral((IntegerLiteralContext) numberCtx).fold();
+            long value = number.longValue();
+            
+            if (value <= 0) {
```

Comment:
does the sequence of 0 timespan make sense at all?

Reply:
I think not - if no maxspan is needed I opted for not declaring it.
If we agree to have `0` indicate no limit across EQL, the parser can be improved.